specs/rollup-node.md

```diff
@@ -373,43 +373,41 @@ The [previous section on L2 chain derivation][l2-chain-derivation] assumes linea
-The starting point of the re-derivation is a pair `(refL2, nextRefL1)` where `refL2` refers to the L2 block to build
-upon and `nextRefL1` refers to the next L1 block to derive from (i.e. if `refL2` is derived from L1 block `refL1`,
-`nextRefL1` is the canonical L1 block at height `l1Number(refL1) + 1`).
-
-In practice, the happy path (no re-org) and the re-org paths are merged. The happy path is simply a special case of the
-re-org path where the starting point of the re-derivation is `(currentL2Head, newL1Block)`.
-
-After a `(currentL2Head, newL1Block)` starting point is found, derivation can continue when a complete sequencing window
-of canonical L1 blocks following the starting point is retrieved.
-
-This re-derivation starting point can be found by applying the following algorithm:
-
-1. (Initialization) Set the initial `refL2` to the head block of the L2 execution engine.
-2. Set `parentL2` to `refL2`'s parent block and `refL1` to the L1 block that `refL2` was derived from.
-3. Fetch `currentL1`, the canonical L1 block at the same height as `refL1`.
-
-- If `currentL1 == refL1`, then `refL2` was built on a canonical L1 block:
-  - Find the next L1 block (it may not exist yet) and return `(refL2, nextRefL1)` as the starting point of the
-    re-derivation.
-    - It is necessary to ensure that no L1 re-org occurred during this lookup, i.e. that `nextRefL1.parent == refL1`.
-    - If the next L1 block does not exist yet, there is no re-org, and nothing new to derive, and we can abort the
-        process.
-- Otherwise, if `refL2` is the L2 genesis block, we have re-orged past the genesis block, which is an error that
-  requires a re-genesis of the L2 chain to fix (i.e. creating a new genesis configuration) (\*)
-- Otherwise, if either `currentL1` does not exist, or `currentL1 != refL1`, set `refL2` to `parentL2` and restart this
-  algorithm from step 2.
-  - Note: if `currentL1` does not exist, it means we are in a re-org to a shorter L1 chain.
-  - Note: as an optimization, we can cache `currentL1` and reuse it as the next value of `nextRefL1` to avoid an
-        extra lookup.
+ancestor, and can re-derive the L2 chain from that L1 block and onward.
+
+The rollup node maintains two heads of the L2 Chain: the unsafe head (often called head) and the safe head.
+Each L2 block has an L1 Attributes block that it references in the L1 Info deposit transaction.
+The unsafe head is the head of the L2 chain. It's L1 Attributes should be canonical or potentially extending the
+canonical chain (if the rollup node has not yet seen the L1 block that it is based upon).
+The safe head is the the last L2 block of the last epoch whose sequencing window is complete
+(i.e. the epoch with number `L1Head.number` - `SEQUENCING_WINDOW_SIZE`).
+
+Determining the unsafe head in the case of an L1 reorg is easy. Start with the current L2 Head and walk the L1 chain
+until the L1 Attributes of an L2 block is canonical or not found (and ahead of the current L1 head).
```

Comment:
Spec writer note: never write "is easy", there is always a chance (small or not) of an oversight in the spec itself.

> until the L1 Attributes of an L2 block is canonical or not found (and ahead of the current L1 head).

Walking back until the L1 attributes are not-found is wrong, the "ahead of the current L1 head" is critical in that case. E.g. you could have a L1-attributes that cannot be found, that builds on top of a L1 chain that is not canonical.